### PR TITLE
Fixes running VSIX on 2015

### LINKED ISSUE #439: Fix VSIX running on VS2015
Due to #430 's addition of CPS references, the current solution does not work on VS2015 and throws runtime errors.

### PR DESCRIPTION
**Pull request by conniey**
* Splits the VSIX project into 2015 and 2017 projects for components that are specific to them. (ie. Common Project System and IThreadingService/IProjectThreadingService)
* Moves common VSIX logic into `ApiPort.VisualStudio.Common` class

Fixes #439

/cc @twsouthwick

### CODE REVIEW

File: src/ApiPort.VisualStudio.2015/ProjectBuilder.cs [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace ApiPortVS.VS2015
+{
+    public class ProjectBuilder : DefaultProjectBuilder
```
Comment (twsouthwick) [inline src/ApiPort.VisualStudio.2015/ProjectBuilder.cs]: Maybe name the type a bit more consie - ProjectBuilder2015 or something

File: src/ApiPort.VisualStudio.2015/ServiceProvider.cs [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+
+            builder.RegisterAdapter<ProjectService, IThreadHandling>(service => service.Services.ThreadingPolicy);
+            builder.RegisterType<VSThreadingService>()
+                .As<IVSThreadingService>()
+                .SingleInstance();
+            builder.RegisterType<ProjectBuilder>()
+                .As<IProjectBuilder>()
+                .SingleInstance();
+
+            base.Load(builder);
```
Comment (twsouthwick) [inline src/ApiPort.VisualStudio.2015/ServiceProvider.cs]: this is a noop

File: src/ApiPort.VisualStudio.2015/VSThreadingService.cs [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+        private readonly IThreadHandling _threadHandling;
+
+        public VSThreadingService(IThreadHandling threadHandling)
+        {
+            _threadHandling = threadHandling;
+        }
+
+        public async Task SwitchToMainThreadAsync(CancellationToken cancellationToken = default(CancellationToken))
+        {
+            await _threadHandling.AsyncPump.SwitchToMainThreadAsync(cancellationToken);
```
Comment (twsouthwick) [inline src/ApiPort.VisualStudio.2015/VSThreadingService.cs]: no need to actually await it - can just return the task

File: src/ApiPort.VisualStudio.2017/ProjectBuilder.cs [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+using System.Diagnostics;
+using Microsoft.VisualStudio.ProjectSystem;
+using Microsoft.VisualStudio.ProjectSystem.Properties;
+using Microsoft.VisualStudio;
+using System.Collections.Concurrent;
+using Microsoft.VisualStudio.ProjectSystem.Build;
+
+namespace ApiPortVS.VS2017
+{
+    public class ProjectBuilder : DefaultProjectBuilder
```
Comment (twsouthwick) [inline src/ApiPort.VisualStudio.2017/ProjectBuilder.cs]: More descriptive name?

File: src/ApiPort.VisualStudio.Common/Analyze/ApiPortVsAnalyzer.cs [hunk 5]
```diff
@@ -60,7 +61,8 @@ public class ApiPortVsAnalyzer : IVsApiPortAnalyzer
             {
                 var issues = _reporter.Issues.ToArray();
 
-                await VisualStudio.ThreadHelper.JoinableTaskFactory.SwitchToMainThreadAsync();
+                //await VisualStudio.ThreadHelper.JoinableTaskFactory.SwitchToMainThreadAsync();
```
Comment (twsouthwick) [inline src/ApiPort.VisualStudio.Common/Analyze/ApiPortVsAnalyzer.cs]: Remove?

File: src/ApiPort.VisualStudio.Common/Analyze/ApiPortVsAnalyzer.cs [hunk 6]
```diff
@@ -81,7 +83,8 @@ public class ApiPortVsAnalyzer : IVsApiPortAnalyzer
             {
                 if (invalidPlatform.Versions.Any(v => v.IsSelected))
                 {
-                    await VisualStudio.ThreadHelper.JoinableTaskFactory.SwitchToMainThreadAsync();
+                    //await VisualStudio.ThreadHelper.JoinableTaskFactory.SwitchToMainThreadAsync();
```
Comment (twsouthwick) [inline src/ApiPort.VisualStudio.Common/Analyze/ApiPortVsAnalyzer.cs]: remove? (there look to be more, but I won't comment on each)

File: src/ApiPort.VisualStudio/ServiceProvider.cs [hunk 7]
```diff
@@ -150,6 +123,19 @@ public ServiceProvider(ApiPortVSPackage serviceProvider)
                 .As<ISourceLineMapper>()
                 .InstancePerLifetimeScope();
 
+            var dte = Package.GetGlobalService(typeof(DTE)) as DTE;
+            var version = new Version(dte.Version);
+
+            switch (version.Major)
```
Comment (twsouthwick) [inline src/ApiPort.VisualStudio/ServiceProvider.cs]: can this just be with an if?

File: src/ApiPort.VisualStudio.2015/VSThreadingService.cs (hunk 3 above)
Comment (conniey) [inline src/ApiPort.VisualStudio.2015/VSThreadingService.cs]: I tried to replace this but the method returns another type `MainThreadAwaitable`, which only has a `GetAwaiter()` method

File: src/ApiPort.VisualStudio.2015/ProjectBuilder.cs (hunk 1 above)
Comment (conniey) [inline src/ApiPort.VisualStudio.2015/ProjectBuilder.cs]: Fixed

File: src/ApiPort.VisualStudio.2015/ServiceProvider.cs (hunk 2 above)
Comment (conniey) [inline src/ApiPort.VisualStudio.2015/ServiceProvider.cs]: Fixed

File: src/ApiPort.VisualStudio.2017/ProjectBuilder.cs (hunk 4 above)
Comment (conniey) [inline src/ApiPort.VisualStudio.2017/ProjectBuilder.cs]: Fixed.. I had changed it to ProjectBuilder2017? :D

File: src/ApiPort.VisualStudio.Common/Analyze/ApiPortVsAnalyzer.cs (hunk 5 above)
Comment (conniey) [inline src/ApiPort.VisualStudio.Common/Analyze/ApiPortVsAnalyzer.cs]: Fixed

File: src/ApiPort.VisualStudio.Common/Analyze/ApiPortVsAnalyzer.cs (hunk 6 above)
Comment (conniey) [inline src/ApiPort.VisualStudio.Common/Analyze/ApiPortVsAnalyzer.cs]: Fixed

File: src/ApiPort.VisualStudio/ServiceProvider.cs (hunk 7 above)
Comment (conniey) [inline src/ApiPort.VisualStudio/ServiceProvider.cs]: Fixed

**Comment by conniey** [comment]
@dotnet-bot test this please

**Pull request merged**